### PR TITLE
Warnings for uninitialized fields in constructors

### PR DESCRIPTION
**Pull request by venkataram-nv**
Addresses #4578 

Adds new tests in the file `tests/diagnostics/uninitialized-fields.slang`.

Warnings will look like the following:

```
.\tests\diagnostics\uninitialized-fields.slang(22): warning 41020: exiting constructor without initializing field 'next'
    __init(int v, Simple* ptr)
    ^~~~~~
.\tests\diagnostics\uninitialized-fields.slang(39): warning 41020: exiting constructor without initializing field 'another_field'
            return;
            ^~~~~~
.\tests\diagnostics\uninitialized-fields.slang(60): warning 41021: default initializer for 'Implicit' will not initialize field 'a'
    int a;
        ^
.\tests\diagnostics\uninitialized-fields.slang(63): warning 41021: default initializer for 'Implicit' will not initialize field 'c'
    int c;
        ^
```

New decorations:
* ConstructorDecoration: indicates whether an IRFunc is a constructor
* SynthesizedDecoration: indicates whether an IRFunc was synthesized (for improved diagnostics)

### CODE REVIEW

File: tests/diagnostics/uninitialized-fields.slang [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+    __init()
+    {
+        field = 0;
+        if (flag == 0)
+            //CHK-DAG: warning 41020: exiting constructor without initializing field 'another_field'
+            return;
+
+        another_field = 0;
+    }
```
Comment (jkwak-work) [inline tests/diagnostics/uninitialized-fields.slang:42]: I am not sure if a return is inserted at the end of the constructor.
Maybe we should have another case to test it like this,
```
    __init()
    {
        field = 0;
        if (flag == 0)
        {
            another_field = 0;
            return;
        }

        //CHK-DAG: warning 41020: exiting constructor without initializing field 'another_field'
    }
```

Comment (jkwak-work) [reply tests/diagnostics/uninitialized-fields.slang:42]: I am not sure if the following case will print a warning,
```
    __init()
    {
        field = 0;
        if (flag == 0)
        {
            another_field = 0;
        }

        return;
    }
```

File: source/slang/slang-ir-use-uninitialized-values.cpp [hunk 2]
```diff
@@ -348,8 +348,135 @@ namespace Slang
...
+            }
+
+        };
+
+        // Work backwards, get exit points and find sources
+        for (auto block : func->getBlocks())
+        {
+            for (auto inst = block->getFirstInst(); inst; inst = inst->next)
+            {
+                auto ret = as<IRReturn>(inst);
```
Comment (jkwak-work) [inline source/slang/slang-ir-use-uninitialized-values.cpp:466]: You may want to use `TerminatorInst` like
```
auto ret = as<TerminatorInst>(inst);
```

File: source/slang/slang-ir-use-uninitialized-values.cpp [hunk 3]
```diff
@@ -348,8 +348,135 @@ namespace Slang
...
+
+    static void checkConstructor(IRFunc* func, ReachabilityContext& reachability, DiagnosticSink* sink)
+    {
+        IRStructType* stype = as<IRStructType>(func->getResultType());
+        if (!stype)
+            return;
+        
+        auto synthesized = func->findDecoration<IRSynthesizedDecoration>();
+        
+        auto emitWarnings = [&](const List<IRStructField*>& fields, IRReturn* ret)
```
Comment (jkwak-work) [inline source/slang/slang-ir-use-uninitialized-values.cpp:440]: I don't think we want to use `emit` for printing warnings.
You may want to reword it to `printWarning`, `printDiagonostics` or something like that.

File: source/slang/slang-ir-use-uninitialized-values.cpp [hunk 4]
```diff
@@ -348,8 +348,135 @@ namespace Slang
...
+    static List<IRStructField*> checkFieldsFromExit(ReachabilityContext& reachability, IRReturn* ret, IRStructType* type)
+    {
+        IRInst* origin = traceInstOrigin(ret->getVal());
+
+        // We don't want to warn on delegated construction
+        if (!isUninitializedValue(origin))
+            return {};
+
+        // Now we can look for all references to fields
+        List<IRStructKey*> usedKeys;
```
Comment (jkwak-work) [inline source/slang/slang-ir-use-uninitialized-values.cpp:404]: You may want to use HashSet type for this variable.
If the user has a struct with hundreds members, it can make a big difference.

**Review by jkwak-work — APPROVED** [review]
Looks good to me.

File: source/slang/slang-ir-use-uninitialized-values.cpp (hunk 4 above)
Comment (venkataram-nv) [reply source/slang/slang-ir-use-uninitialized-values.cpp:404]: Thanks, I was looking for a data structure to use here instead.

File: source/slang/slang-ir-use-uninitialized-values.cpp (hunk 3 above)
Comment (venkataram-nv) [reply source/slang/slang-ir-use-uninitialized-values.cpp:440]: Sounds good. Is lambda use alright?

File: tests/diagnostics/uninitialized-fields.slang (hunk 1 above)
Comment (venkataram-nv) [reply tests/diagnostics/uninitialized-fields.slang:42]: It seems that constructors will always end with a return, so in the first case there will be a warning that is printed. The second case does not issue a warning because we do not checking for potentially uninitialized values. We only trigger warnings if it a field will *never* be initialized by the time of next use/return;

File: tests/compute/struct-default-init.slang [hunk 5]
```diff
@@ -3,9 +3,9 @@
 
 struct Test
 {
-	int a;
```
[the remark this answers is not in the document]
Comment (venkataram-nv) [reply tests/compute/struct-default-init.slang:6]: It currently does indeed issue a warning here. The reason is that a default constructor is still created, and within that default constructor the field `a` is not initialized. I think its worth keeping the warning; even if it is not a problem in the current file, it could be a problem when using with modules.

File: source/slang/slang-ast-decl.h [hunk 6]
```diff
@@ -371,6 +371,8 @@ class FunctionDeclBase : public CallableDecl
 class ConstructorDecl : public FunctionDeclBase
 {
     SLANG_AST_CLASS(ConstructorDecl)
+    
+    bool synthesized = false;
```
[the remark this answers is not in the document]
Comment (venkataram-nv) [reply source/slang/slang-ast-decl.h:375]: Its meant to be like a flag that indicates if a constructor has been synthesized by slangc and not provided by the user. Its used later to handle implicitly created default constructors.

File: source/slang/slang-ir-use-uninitialized-values.cpp (hunk 2 above)
Comment (venkataram-nv) [reply source/slang/slang-ir-use-uninitialized-values.cpp:466]: We need an explicit return so we can get the value returned (constructed) and trace is back so a `IRTerminatorInst` wouldn't be enough.

Comment (jkwak-work) [reply source/slang/slang-ir-use-uninitialized-values.cpp:470]: There are other ways to return from the middle of the constructor.
`__intrinstic_asm` is one of them.
And they are grouped into `IRTerminatorInst`.

File: tests/compute/struct-default-init.slang (hunk 5 above)
[the remark this answers is not in the document]
Comment (jkwak-work) [reply tests/compute/struct-default-init.slang:6]: I am not sure if we want to print a warning about this case,
```
struct Test
{
    int a;
};
```
C/C++ doesn't do that....

File: source/slang/slang-ir-use-uninitialized-values.cpp (hunk 3 above)
Comment (jkwak-work) [reply source/slang/slang-ir-use-uninitialized-values.cpp]: We have a rule for lambda.
```
Lambdas are allowed, but think carefully about whether just declaring a subroutine would also work.
```
https://github.com/jkwak-work/slang/blob/master/docs/design/coding-conventions.md
In general, it is better to stay away from it.
In this case, it is called only once, my preference is not to use it, but it seems fine.

File: source/slang/slang-ast-decl.h [hunk 7]
```diff
@@ -371,6 +371,10 @@ class FunctionDeclBase : public CallableDecl
 class ConstructorDecl : public FunctionDeclBase
 {
     SLANG_AST_CLASS(ConstructorDecl)
+   
+    // Indicates whether the declaration was synthesized by
+    // slang and not actually provided by the user
+    bool isSynthesized = false;
```
Comment (jkwak-work) [inline source/slang/slang-ast-decl.h:377]: I am still little skeptical about the word "synthesize".
It is fine as it is but can you tell me how you would define the word in this context?

File: source/slang/slang-ir-use-uninitialized-values.cpp [hunk 8]
```diff
@@ -348,8 +348,139 @@ namespace Slang
...
     }
 
+    static bool isInstStoredInto(ReachabilityContext& reachability, IRInst* reference, IRInst* inst)
+    {
+        auto addresses = getAliasableInstructions(inst);
+        
+        List<IRInst*> stores;
+        List<IRInst*> loads;
+
+        for (auto alias : addresses)
```
Comment (jkwak-work) [inline source/slang/slang-ir-use-uninitialized-values.cpp:358]: Probably easier to read if this was
```
for (auto alias : getAliasableInstructions(inst))
```

**Review by jkwak-work — APPROVED** [review]
Looks still good to me.

File: source/slang/slang-ast-decl.h (hunk 7 above)
Comment (venkataram-nv) [reply source/slang/slang-ast-decl.h:377]: Synthesize here means a declaration generated by slang which was never actually parsed from the file.

File: source/slang/slang-ir-use-uninitialized-values.cpp (hunk 3 above)
Comment (venkataram-nv) [reply source/slang/slang-ir-use-uninitialized-values.cpp]: I see. In this case I wanted to avoid passing multiple parameters to a separate function.

File: tests/compute/struct-default-init.slang (hunk 5 above)
[the remark this answers is not in the document]
Comment (venkataram-nv) [reply tests/compute/struct-default-init.slang:6]: That is true. I will see if I can make an exception for this.

**Review by jkwak-work — APPROVED** [review]
Looks good.
